src/auth/azureAuth.ts

```diff
@@ -0,0 +1,246 @@
+import {
+    Disposable as VsCodeDisposable,
+    EventEmitter,
+    authentication,
+    AuthenticationGetSessionOptions,
+    AuthenticationSession,
+    window,
+    QuickPickItem,
+} from "vscode";
+import { SignInStatus, TokenInfo } from "./types";
+import { Environment } from "@azure/ms-rest-azure-env";
+import { getConfiguredAzureEnv } from "../commands/utils/config";
+import { Errorable, failed, getErrorMessage, map as errmap, succeeded, bindAsync } from "../commands/utils/errorable";
+import { TokenCredential } from "@azure/core-auth";
+import { parseJson } from "../commands/utils/json";
+import { SubscriptionClient, TenantIdDescription } from "@azure/arm-resources-subscriptions";
+import { listAll } from "../commands/utils/arm";
+
+type AuthProviderId = "microsoft" | "microsoft-sovereign-cloud";
+
+type Tenant = {
+    name: string;
+    id: string;
+};
+
+class AzureSessionProvider extends VsCodeDisposable {
+    private readonly initializePromise: Promise<void>;
+    private tenants: Tenant[] = [];
+
+    public readonly onSignInStatusChangeEmitter = new EventEmitter<SignInStatus>();
+    public signInStatusValue: SignInStatus = "Initializing";
+
+    public constructor() {
+        const disposable = authentication.onDidChangeSessions(async (e) => {
+            // Ignore events for non-microsoft providers
+            if (e.provider.id !== getConfiguredAuthProviderId()) {
+                return;
+            }
+
+            // Ignore events while we're initializing.
+            if (this.signInStatusValue === "Initializing") {
+                return;
+            }
+
+            // Silently check authentication status and tenants
+            await this.signInAndUpdateTenants({ createIfNone: false, silent: true });
+        });
+
+        super(() => {
+            this.onSignInStatusChangeEmitter.dispose();
+            disposable.dispose();
+        });
+
+        this.initializePromise = this.initialize();
+    }
+
+    public get signInStatus(): SignInStatus {
+        return this.signInStatusValue;
+    }
+
+    private async initialize(): Promise<void> {
+        await this.signInAndUpdateTenants({ createIfNone: false, silent: true });
+    }
+
+    public async signIn(): Promise<void> {
+        await this.initializePromise;
+
+        const newSignInStatus = "SigningIn";
+        if (newSignInStatus !== this.signInStatusValue) {
+            this.signInStatusValue = newSignInStatus;
+            this.onSignInStatusChangeEmitter.fire(this.signInStatusValue);
+        }
+        await this.signInAndUpdateTenants({ createIfNone: true, clearSessionPreference: true });
+    }
+
+    private async signInAndUpdateTenants(options: AuthenticationGetSessionOptions): Promise<void> {
```

Comment:
💡 This whole experience needs quick test form folks with non msft account as well, I have tried to give a good shake, things look good to me,

I feel this is perfect opportunity for us to improve how first sign-up will happen. If there is any further enhancement is needed we can improve as well go forward.